lib/rubocop/cop/performance/chain_array_allocation.rb

```diff
@@ -0,0 +1,81 @@
+    module Performance
+      # This cop is used to identify usages of
+      # @example
+      #   # bad
+      #   array = ["a", "b", "c"]
+      #   array.compact.flatten.map { |x| x.downcase }
+      #
+      # Each of these methods (`compact`, `flatten`, `map`) will generate a
+      # new intermediate array that is promptly thrown away. Instead it is
+      # faster to mutate when we know it's safe.
+      #
+      # @example
+      #   # good.
+      #   array = ["a", "b", "c"]
+      #   array.compact!
+      #   array.flatten!
+      #   array.map! { |x| x.downcase }
+      #   array
+      class ChainArrayAllocation < Cop
+        include RangeHelp
+
+        # These methods return a new array but only sometimes. They must be
+        # called with an argument. For example:
+        #
+        #   [1,2].first    # => 1
+        #   [1,2].first(1) # => [1]
+        #
+        RETURN_NEW_ARRAY_WHEN_ARGS = ':first :last :pop :sample :shift '.freeze
+
+        # These methods return a new array only when called without a block.
+        RETURNS_NEW_ARRAY_WHEN_NO_BLOCK = ':zip :product '.freeze
+
+        # These methods ALWAYS return a new array
+        # after they're called it's safe to mutate the the resulting array
+        ALWAYS_RETURNS_NEW_ARRAY = ':* :+ :- :collect :compact :drop '\
+                                   ':drop_while :flatten :map :reject ' \
+                                   ':reverse :rotate :select :shuffle :sort ' \
+                                   ':take :take_while :transpose :uniq ' \
+                                   ':values_at :| '.freeze
+
+        # These methods have a mutation alternative. For example :collect
+        # can be called as :collect!
+        HAS_MUTATION_ALTERNATIVE = ':collect :compact :flatten :map :reject '\
+                                   ':reverse :rotate :select :shuffle :sort '\
+                                   ':uniq '.freeze
+        MSG = 'Use `%<method>s...%<second_method>s!` instead of `%<method>s' \
+              '...%<second_method>s`.'.freeze
+
+        def_node_matcher :flat_map_candidate?, <<-PATTERN
+          {
+            (send (send _ ${#{RETURN_NEW_ARRAY_WHEN_ARGS}} {int lvar ivar cvar gvar}) ${#{HAS_MUTATION_ALTERNATIVE}} $...)
+            (send (block (send _ ${#{ALWAYS_RETURNS_NEW_ARRAY} }) ...) ${#{HAS_MUTATION_ALTERNATIVE}} $...)
+            (send (send _ ${#{ALWAYS_RETURNS_NEW_ARRAY + RETURNS_NEW_ARRAY_WHEN_NO_BLOCK}} ...) ${#{HAS_MUTATION_ALTERNATIVE}} $...)
+          }
+        PATTERN
+
+        def on_send(node)
+          flat_map_candidate?(node) do |fm, sm, _|
+            range = range_between(
+              node.loc.dot.begin_pos,
+              node.source_range.end_pos
+            )
+            add_offense(
+              node,
+              location: range,
+              message: format(MSG, method: fm, second_method: sm)
+            )
+          end
+        end
+
+        def autocorrect(node)
+          # ...
+        end
```

Comment:
When the `autocorrect ` method is defined, `rake generate_cops_documentation` task generates a document that shows this cop supports auto-correct.

```markdown
## Performance/ChainArrayAllocation

Enabled by default | Supports autocorrection
--- | ---
Disabled | Yes
```

https://github.com/rubocop-hq/rubocop/pull/6234/files#diff-adc5037eb2a271f411b1219b8d593467

It is good practice to remove this `autocorrect` method definition until it supports auto-correct.

Reply:
Yeah, that's a good point.